lucid/modelzoo/get_activations.py

```diff
@@ -0,0 +1,170 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Helpers for getting responses from models over large collections."""
+
+
+import itertools
+from collections import defaultdict
+
+import numpy as np
+import tensorflow as tf
+
+from lucid.misc.iter_nd_utils import recursive_enumerate_nd, dict_to_ndarray, batch_iter
+
+
+def get_activations_iter(model, layer, generator, reducer="mean", batch_size=64,
+                         dtype=None, ind_shape=None, center_only=False):
+  """Collect center activtions of a layer over many images from an iterable obj.
+
+  Note: this is mostly intended for large synthetic families of images, where
+    you can cheaply generate them in Python. For collecting activations over,
+    say, ImageNet, there will be better workflows based on various dataset APIs
+    in TensorFlow.
+
+  Args:
+    model: model for activations to be collected from.
+    layer: layer (in model) for activtions to be collected from.
+    generator: An iterable object (intended to be a generator) which produces
+      tuples of the form (index, image). See details below.
+    reducer: How to combine activations if multiple images map to the same index.
+      Supports "mean", "rms", and "max".
+    batch_size: How many images from the generator should be processes at once?
+    dtype: determines dtype of returned data (defaults to model activation
+      dtype). Can be used to make funciton memory efficient.
+    ind_shape: Shape that indices can span. Optional, but makes funciton orders
+      of magnitiude more memory efficient.
+
+  Memory efficeincy:
+    Using ind_shape is the main tool for make this function memory efficient.
+    dtype="float16" can further help.
+
+  Returns:
+    A numpy array of shape [ind1, ind2, ..., layer_channels]
+  """
+
+
+  assert reducer in ["mean", "max", "rms"]
+  combiner, normalizer = {
+      "mean" : (lambda a,b: a+b,             lambda a,n: a/n         ),
+      "rms"  : (lambda a,b: a+b**2,          lambda a,n: np.sqrt(a/n)),
+      "max"  : (lambda a,b: np.maximum(a,b), lambda a,n: a           ),
+  }[reducer]
+
+  with tf.Graph().as_default(), tf.Session() as sess:
+    t_img = tf.placeholder("float32", [None, None, None, 3])
+    T = model.import_graph(t_img)
+    t_layer = T(layer)
+
+    responses = None
+    count = None
+
+    # # If we know the total length, let's give a progress bar
+    # if ind_shape is not None:
+    #   total = int(np.prod(ind_shape))
+    #   generator = tqdm(generator, total=total)
+
+    for batch in batch_iter(generator, batch_size=batch_size):
+
+      inds, imgs = [x[0] for x in batch], [x[1] for x in batch]
+
+      # Get activations (middle of image)
+      acts = t_layer.eval({t_img: imgs})
+      if center_only:
+        acts = acts[:, acts.shape[1]//2, acts.shape[2]//2]
```

Comment:
would it be worth building in support for `NCHW` formatted activations through a `is_nchw=False` flag? the multimodal model uses NCHW. Another option is to do an NCHW transpose as part of `T()` automatically with an opt-out.

Reply:
Ideally, I'd like to use `lucid.misc.graph_analysis.infer_format()` here so that it's automatic, but don't fully trust it yet. Will defer for now.

Reply:
In the original `get_responses()` implementation, in case the spatial dimensions had even length, we would average over the center 4 points. I think we want that behavior in this implementation, too? Here's the original code that implemented it; may need to be special cased for very small shapes:
```
D = (acts_.shape[0]+1)//2 - 1
        if D > 0:
          acts_ = acts_[D:-D,D:-D]
        resp = acts_.max(0).max(0)[unit_list]
```
(Ludwig via Nick's account; we were debugging the slight difference between activations when passing `center_only=True` to this.)